server/src/main/java/org/elasticsearch/index/mapper/MetadataFieldMapper.java

```diff
@@ -45,6 +45,14 @@
          * @param parserContext context that may be useful to build the field like analyzers
          */
         MetadataFieldMapper getDefault(ParserContext parserContext);
+
+        /**
+         *  @return Whether a metadata field can be included in the document _source.
+         */
+        default boolean isAllowedInSource() {
```

Comment:
I'm wondering if we really need this new flag `isAllowedInSource`. Maybe we could avoid adding a new flag and instead do the following:
* For metadata fields that are not allowed in _source, make sure that `MetadataFieldMapper#parse` throws an descriptive error.
* Many metadata fields currently have logic in `parse` that's unrelated to _source parsing. We could make sure to move it into the special methods `preParse` or `postParse`.

Reply:
@jtibshirani I agree with you that delegating this to `MetadataFieldMapper#parse()` would be a simpler solution. However I see two caveats: 

- The approach with `isAllowedInSource()` method detects that a metadata field is being parsed when parsing the field name. This is early in the parsing process. An  exception is thrown and the field value parsing is skipped. 

- As you mention, I have seen classes (such as `IgnoredFieldMapper` and `VersionFieldMapper`) that override `parse()` method to do nothing on one hand, while calling `super.parse()` from `preParse()`/`postParse()` methods. This approach is too complicated imho, but I am not sure that refactoring this is very simple. WDYT?

Reply:
> This is early in the parsing process. An exception is thrown and the field value parsing is skipped.

I don't think we need to optimize performance in this case because it's an error condition (and should be a rare one too).

> This approach is too complicated imho, but I am not sure that refactoring this is very simple.

I agree that the logic is complex/ hard to read! I haven't deeply looked into the refactoring myself -- maybe you could try it out and we can rediscuss the approach if you see a roadblock ?